packages/errors/lib/http-error.js

```diff
@@ -0,0 +1,152 @@
+/**
+ * @module @dotcom-reliability-kit/errors/lib/http-error
+ */
+
+const OperationalError = require('./operational-error');
+
+/**
+ * We have guards in place wherever we use this map of status messages
+ * which means we can safely cast it to an object where every property
+ * is a string. If we don't do this then TypeScript will complain.
+ * {@see HttpError.getMessageForStatusCode}
+ *
+ * @type {Object<String, String>}
+ */
+const STATUS_CODES = require('http').STATUS_CODES;
+
+/**
+ * Class representing an HTTP error.
+ */
+class HttpError extends OperationalError {
+	/**
+	 * @readonly
+	 * @access public
+	 * @type {String}
+	 */
+	name = 'HttpError';
+
+	/**
+	 * @readonly
+	 * @access public
+	 * @type {Number}
+	 */
+	statusCode = 500;
+
+	/**
+	 * @readonly
+	 * @access public
+	 * @type {String}
+	 */
+	statusMessage = STATUS_CODES[500];
+
+	/**
+	 * @readonly
+	 * @access public
+	 * @type {Number}
+	 */
+	get status() {
+		return this.statusCode;
+	}
+
+	/**
+	 * Create an HTTP error.
+	 *
+	 * @param {(String|Number|HttpErrorData)} [data = {}]
+	 *     The error message if it's a string, the HTTP status code if it's a number, or full error
+	 *     information if an object.
+	 */
+	constructor(data = {}) {
+		if (typeof data === 'string') {
+			data = { message: data };
+		}
+		if (typeof data === 'number') {
+			data = { statusCode: data };
+		}
+
+		// Make sure that we don't modify the original data object
+		// by shallow-cloning it
+		data = { ...data };
+
+		// Default the status code
+		data.statusCode =
+			typeof data.statusCode === 'number'
+				? HttpError.normalizeErrorStatusCode(data.statusCode)
+				: 500;
+
+		// Default the error code
+		data.code =
+			typeof data.code === 'string' ? data.code : `HTTP_${data.statusCode}`;
+
+		// Default the error message
+		data.message =
+			typeof data.message === 'string'
+				? data.message
+				: HttpError.getMessageForStatusCode(data.statusCode);
+
+		// Message and code are set by the parent class, but we need
+		// to set the status code and message properties ourselves
+		super(data);
```

Comment:
**suggestion**: rather than cloning and overriding the data object, you could construct a new object to pass to `super` (since you're setting every property anyway)? would probably be cleaner

Reply:
Ah! Yes I could, it's obvious now I see it 😄  thanks

Reply:
Ooh but actually if we define a new object, it needs to have _all_ of the properties from `data` on it anyway because `OperationalError` collects those and adds them to the `data` property. So we end up with something like this which is essentially just renaming what we already have. Maybe that's still better?

```js
// Define a new data object so that we don't modify the original
const sanitizedData = { ...data };

// Default the status code
sanitizedData.statusCode =
	typeof data.statusCode === 'number'
		? HttpError.normalizeErrorStatusCode(data.statusCode)
		: 500;

// Default the error code
sanitizedData.code =
	typeof data.code === 'string'
		? data.code
		: `HTTP_${sanitizedData.statusCode}`;

// Default the error message
sanitizedData.message =
	typeof data.message === 'string'
		? data.message
		: HttpError.getMessageForStatusCode(sanitizedData.statusCode);

// Message and code are set by the parent class, but we need
// to set the status code and message properties ourselves
super(sanitizedData);
this.statusCode = sanitizedData.statusCode;
this.statusMessage = HttpError.getMessageForStatusCode(
	sanitizedData.statusCode
);
```

Open to other suggestions!

Reply:
oh, yeah, sure, that's no better really is it